Update support story for external calls and add sample (#204)
OfficeDev/office-scripts-docs: docs/resources/scenarios/noaa-data-fetch.md

---
title: 'Office Scripts sample scenario: Graph water-level data from NOAA'
description: 'A sample that fetches JSON data from a NOAA database and uses it to create a chart.'
ms.date: 12/28/2020
localization_priority: Normal
---

# Office Scripts sample scenario: Graph water-level data from NOAA

In this scenario, you need to plot the water level at the [National Oceanic and Atmospheric Administration's Seattle station](https://tidesandcurrents.noaa.gov/stationhome.html?id=9447130). You'll use external data to populate a spreadsheet and create a chart.

You'll develop a script that uses the `fetch` command to query the [NOAA Tides and Currents database](https://tidesandcurrents.noaa.gov/). That will get the water level recorded across a given time span. The information will be returned as JSON, so part of the script will translate that into range values. Once the data is in the spreadsheet, it will be used to make a chart.

## Scripting skills covered

- External API calls (`fetch`)
- JSON parsing
- Charts

## Setup instructions

> 1. Open the workbook with Excel for the web.

Reviewer: ```suggestion
1. Open the workbook with Excel on the web.
```

> 1. Under the **Automate** tab, open **All Scripts**.

Reviewer: Since **All Scripts** is a button and menu item, per the [style guide](https://styleguides.azurewebsites.net/Styleguide/Read?id=2700&topicid=26472) the verb should be 'select'. 
```suggestion
1. Under the **Automate** tab, select **All Scripts**.
```

> 1. In the **Code Editor** task pane, press **New Script** and paste the following script into the editor.

Reviewer: Also a button. Another reason not to use 'press' is that it is less inclusive -- it is mouse-centric and implies one is not using keyboard or voice navigation. 
```suggestion
1. In the **Code Editor** task pane, select **New Script** and paste the following script into the editor.
```

      const option = "water_level";
      const startDate = "20201225"; /* yyyymmdd date format */
      const endDate = "20201227";
      const station = "9447130"; /* Seattle */
    
      // Construct the URL for the fetch call.
      const strQuery = `https://api.tidesandcurrents.noaa.gov/api/prod/datagetter?product=${option}&begin_date=${startDate}&end_date=${endDate}&datum=MLLW&station=${station}&units=english&time_zone=gmt&application=NOS.COOPS.TAC.WL&format=json`;
    
      console.log(strQuery);
    
      // Resolve the Promises returned by the fetch operation.
      const response = await fetch(strQuery);
      const rawJson = await response.json();
    
      // Translate the raw JSON into a usable state.
      const stringifiedJson = JSON.stringify(rawJson);
      const noaaData = JSON.parse(stringifiedJson);
    
      // Create table headers and format them to stand out.
      let headers = [["Time", "Level"]];
      let headerRange = currentSheet.getRange("A1:B1");
      headerRange.setValues(headers);
      headerRange.getFormat().getFill().setColor("#4472C4");
      headerRange.getFormat().getFont().setColor("white");
    

>       // Insert all the data in rows from JSON

Reviewer: ```suggestion
      // Insert all the data in rows from JSON.
```

      let noaaDataCount = noaaData.data.length;
      let dataToEnter = [[], []]
      for (let i = 0; i < noaaDataCount; i++) {
        let currentDataPiece = noaaData.data[i];
        dataToEnter[i] = [currentDataPiece.t, currentDataPiece.v];
      }
    
      let dataRange = currentSheet.getRange("A2:B" + String(noaaDataCount + 1)); /* +1 to account for the title row */
      dataRange.setValues(dataToEnter);
      
      // Format the "Time" column for timestamps.
      dataRange.getColumn(0).setNumberFormatLocal("[$-en-US]mm/dd/yyyy hh:mm AM/PM;@");
    
      // Create and format a chart with the level data.
      let chart = currentSheet.addChart(ExcelScript.ChartType.xyscatterSmooth,dataRange);
      chart.getTitle().setText("Water Level - Seattle");
      chart.setTop(0);
      chart.setLeft(300);
      chart.setWidth(500);
      chart.setHeight(300);
      chart.getAxes().getValueAxis().setShowDisplayUnitLabel(false);
      chart.getAxes().getCategoryAxis().setTextOrientation(60);
      chart.getLegend().setVisible(false);
    }
    ```

1. Rename the script to **NOAA Water Level Chart** and save it.

## Running the script

> On any worksheet, run the **NOAA Water Level Chart** script. The script fetches the water level data from December 25th 2020 to December 27th 2020. The `const` variables at the beginning of the script can be changed to use different dates or get different station information. The [CO-OPS API For Data Retrieval](https://api.tidesandcurrents.noaa.gov/api/prod/) describes how to get all this data.

Reviewer: Per [style guide](https://styleguides.azurewebsites.net/Styleguide/Read?id=2700&topicid=26913). 
```suggestion
On any worksheet, run the **NOAA Water Level Chart** script. The script fetches the water level data from December 25, 2020 to December 27, 2020. The `const` variables at the beginning of the script can be changed to use different dates or get different station information. The [CO-OPS API For Data Retrieval](https://api.tidesandcurrents.noaa.gov/api/prod/) describes how to get all this data.
```
